Improve import manipulation of Fortran files

jubich opened the pull request:
This PR adds new scripts for editing/formatting imports in Fortran files (`.F90 and .f90`). 

`sort_modules` now sorts not only the modules alphabetically, but also the imported functions/methods. This script was also used for sorting the imports in #1382. You might consider switching from `.sort()` to .`sort(key=str.casefold)` to get a case-insensitive sort.

`check_import_usage` checks whether the imported functions/methods (with the exception of assignment() and operator()) are used or not. This is done by finding all "import lines". These are removed and then the remaining file is converted into a `set` of lowercase words. These are then compared with a `set` created from the imported functions/methods.

aradi reviewed (commented):
Please adapt to line lenght 100. (You can reformat mit `black -l 100`)

aradi commented:
ping (unstale)

aradi commented on utils/srcmanip/check_import_usage:
```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env python3
+
+"""Simple script to ckeck usage of imports in Fortran files"""
```
```suggestion
"""Simple script to check usage of imports in Fortran files"""
```

aradi commented on utils/srcmanip/check_import_usage:
```diff
@@ -0,0 +1,110 @@
+    (?P<name>\w+)
+    (?P<rest>.*?(?:&[ ]*\n(?:[ ]*&)?.*?)*)\n
+    """,
+    re.VERBOSE | re.MULTILINE,
```
Maybe we should add also `re.IGNORECASE` (as we're dealing with Fortran :wink: )

aradi commented on utils/srcmanip/check_import_usage:
```diff
@@ -0,0 +1,110 @@
+            for match in matches:
+                if match["rest"]:
+                    rest_list.append(match["rest"])
+                file_txt = file_txt[: match.start()] + file_txt[match.end() :]
```
What's the point to check the source *before* the import? It is enough to check only the text which comes after the match.

aradi replied:
Could, you, please also try to rebase it to current main?

aradi reviewed (changes requested):
Thanks for the PR. Please change following:
* remove the `--files` and `--folders` options in both scripts, that is quite unusual. The commands should simply accept files as positional arguments. (We can easily generate file names in folders, if necessary).
* We should add a warning into the `check_import_usage` that it uses a very simple heuristics, so that people are aware, that it can fooled in many ways.

jubich commented:
I have now changed both files so that they now accept both **positional** and optional arguments. This allows the scripts to be called in the _usual_ way with positional arguments as well as _unusually_ with optional arguments. This also allows to call both functions e.g. without files, but only with folders.


I have also “improved” the import checker so that comments are ignored. Regarding your question, why lines before the import should be checked:
They obviously do not need to be checked. The current implementation first searches _all_ imports. Then it reverses the order, removes the matches from the read file and stores the matches in a `list`. The reversal is necessary because otherwise the `.start()` and `.end()` positions of the matches would be wrong due to the shift after removing the above import. I now additionally remove the lines before the first import (which mainly removes the header). 
Next, the comments are removed and the remaining file is converted to a `set`.
Now the `list` with the saved imports is also converted into a `set` and the difference between the two `set's` is calculated. The difference then contains the _unused_ imports. 

I cannot simply remove all lines before the last import, because sometimes there are imports that are not at the beginning of the file. So I wouldn't check the usage between such imports and imports at the beginning of the file. I could check the file for each import individually, but I doubt that would bring much benefit.
